README.md
![Logo](banner.jpg)

# This is a very early fork, recommend using the original for now.





<p align="center">
    <a href="https://github.com/TheAnkurPanchani/flutter_swiper/pulls">
        <img src="https://img.shields.io/badge/PRs-Welcome-brightgreen.svg" alt="PRs Welcome" />
    </a>
    <a href="https://pub.dev/packages/card_swiper">
        <img src="https://img.shields.io/pub/v/card_swiper.svg" alt="pub package" />
    </a>
</p>

# card_swiper

Swiper/Carousel for flutter, with multiple layouts, infinite loop. Compatible with Android & iOS.

# :sparkles::sparkles: New Features:Layout for pagination.



# :sparkles::sparkles: New Features:PageTransformer

Finally, we have `PageTransformer` like android, just set a `transformer` to `Swiper`,
it returns a widget that has been transformed. For now, only support for layout `DEFAULT`.
Thanks to @FlutterRocks ,you've done great job 👏.



# :sparkles::sparkles: New Features:Layout

![](https://github.com/jzoom/images/raw/master/layout1.gif)

![](https://github.com/jzoom/images/raw/master/layout2.gif)

![](https://github.com/jzoom/images/raw/master/layout3.gif)

[See More](#build-in-layouts)

# Showcases

![Horizontal](https://github.com/jzoom/flutter_swiper/raw/master/example/res/1.gif)

![Vertical](https://github.com/jzoom/flutter_swiper/raw/master/example/res/2.gif)

![Custom Pagination](https://github.com/jzoom/flutter_swiper/raw/master/example/res/3.gif)

![Custom Pagination](https://github.com/jzoom/flutter_swiper/raw/master/example/res/4.gif)

![Phone](https://github.com/jzoom/flutter_swiper/raw/master/example/res/5.gif)

![Example](https://github.com/jzoom/images/raw/master/swiper-example.gif)

[See More](#codes)

## Changelogs

> see:[CHANGELOG.md](https://github.com/TheAnkurPanchani/flutter_swiper/blob/master/CHANGELOG.md)

## Getting Started

- [Installation](#installation)
- [Basic Usage](#basic-usage)
- [Constructor](#constructor)
  - [Basic](#basic)
  - [Pagination](#pagination)
  - [Control buttons](#control-buttons)
  - [Controller](#controller)
  - [Autoplay](#autoplay)
- [Build in layouts](#build-in-layouts)
- [Codes](#codes)

### Installation

Add

```bash
carousel_animations : 
```

to your `pubspec.yaml`, and run

```bash
flutter packages get
```

in your project's root directory.

OR

run

```bash
flutter pub add carousel_animations
```

in your project's root directory.

### Basic Usage

Create a new project with command

```
flutter create myapp
```

Edit lib/main.dart like this:

```dart

import 'package:flutter/material.dart';

import 'package:carousel_animations/carousel_animations.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
    body: Swiper(
        itemBuilder: (BuildContext context,int index){
          return Image.network("https://via.placeholder.com/350x150",fit: BoxFit.fill,);
        },
        itemCount: 3,
        pagination: SwiperPagination(),
        control: SwiperControl(),
      ),
    );
  }
}

```

### Constructor

#### Basic

| Parameter       |            Default             | Description                                                                                                                                                              |
| :-------------- | :----------------------------: | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| scrollDirection |        Axis.horizontal         | If `Axis.horizontal`, the scroll view's children are arranged horizontally in a row instead of vertically in a column.                                                   |
| axisDirection   |       AxisDirection.left       | If `AxisDirection.right`, the scroll view's children are arranged right side in a row instead of left.<br>**Note:** Currently supported for horizontal stack layout only |
| loop            |              true              | Set to `false` to disable continuous loop mode.                                                                                                                          |
| index           |               0                | Index number of initial slide.                                                                                                                                           |
| autoplay        |             false              | Set to `true` enable auto play mode.                                                                                                                                     |
| onIndexChanged  | void onIndexChanged(int index) | Called with the new index when the user swiped or autoplay                                                                                                               |
| onTap           |     void onTap(int index)      | Called when user tap ui.                                                                                                                                                 |
| duration        |             300.0              | The milliscends of every transaction animation costs                                                                                                                     |
| pagination      |              null              | set `SwiperPagination()` to show default pagination                                                                                                                      |
| control         |              null              | set `SwiperControl()` to show default control buttons                                                                                                                    |

#### Pagination

The pagination extends from `SwiperPlugin`,the `SwiperPlugin` provides extra ui for `Swiper`.Set `SwiperPagination()` to show default pagination.

| Parameter |          Default           | Description                                                                                                                               |
| :-------- | :------------------------: | :---------------------------------------------------------------------------------------------------------------------------------------- |
| alignment |   Alignment.bottomCenter   | Change this value if you what to put pagination in other place                                                                            |
| margin    | const EdgeInsets.all(10.0) | The distance between inner side of the parent container.                                                                                  |
| builder   |   SwiperPagination.dots    | There are three default styles `SwiperPagination.dots`, `SwiperPagination.fraction` and `SwiperPagination.rect`, these can be customized. |

If you'd like to customize your own pagination, you can do like this:

```dart
Swiper(
    ...,
    pagination: SwiperCustomPagination(
        builder:(BuildContext context, SwiperPluginConfig config){
            return YourOwnPagination();
        }
    )
);

```

##### Implemented pagination

###### DotSwiperPaginationBuilder

| Parameter   | Default                                   | Description              | Required |
| :---------- | :---------------------------------------- | :----------------------- | :------: |
| activeColor | Theme.of(context).primaryColor            | Active bullet color      |  false   |
| color       | Theme.of(context).scaffoldBackgroundColor | Bullet color             |  false   |
| activeSize  | 10.0                                      | Active bullet size       |  false   |
| size        | 10.0                                      | Bullet size              |  false   |
| space       | 3.0                                       | Distance between bullets |  false   |
| key         | -                                         | key                      |  false   |

###### FractionPaginationBuilder

| Parameter      | Default                                   | Description       | Required |
| :------------- | :---------------------------------------- | :---------------- | :------: |
| activeColor    | Theme.of(context).primaryColor            | Active font color |  false   |
| color          | Theme.of(context).scaffoldBackgroundColor | font color        |  false   |
| activeFontSize | 35.0                                      | Active font size  |  false   |
| fontSize       | 20.0                                      | Font size         |  false   |
| key            | -                                         | key               |  false   |

###### RectSwiperPaginationBuilder

| Parameter   | Default                                   | Description              | Required |
| :---------- | :---------------------------------------- | :----------------------- | :------: |
| activeColor | Theme.of(context).primaryColor            | Active bullet color      |  false   |
| color       | Theme.of(context).scaffoldBackgroundColor | Bullet color             |  false   |
| activeSize  | 10.0                                      | Active bullet size       |  false   |
| size        | 10.0                                      | Bullet size              |  false   |
| space       | 3.0                                       | Distance between bullets |  false   |
| key         | -                                         | key                      |  false   |

#### Control buttons

The control also extends from `SwiperPlugin`,set `SwiperControl()` to show default control buttons.

| Parameter    |             Default             | Description                                        |
| :----------- | :-----------------------------: | :------------------------------------------------- |
| iconPrevious |      Icons.arrow_back_ios       | The icon data to display `previous` control button |
| iconNext     |     Icons.arrow_forward_ios     | The icon data to display `next`.                   |
| color        | Theme.of(context).primaryColor  | Control button color                               |
| disableColor | Theme.of(context).disabledColor | Disabled control button color                      |
| size         |              30.0               | Control button size                                |
| padding      |    const EdgeInsets.all(5.0)    | Control button padding                             |

#### Controller

The `Controller` is used to control the `index` of the Swiper, start or stop autoplay.You can create a controller by `SwiperController()` and save the instance by futher usage.

| Method                                       | Description                                         |
| :------------------------------------------- | :-------------------------------------------------- |
| void move(int index, {bool animation: true}) | Move to the spicified `index`,with animation or not |
| void next({bool animation: true})            | Move to next                                        |
| void previous({bool animation: true})        | Move to previous                                    |
| void startAutoplay()                         | Start autoplay                                      |
| void stopAutoplay()                          | Stop autoplay                                       |

#### Autoplay

| Parameter                    | Default | Description                                          |
| :--------------------------- | :-----: | :--------------------------------------------------- |
| autoplayDelay                |  3000   | Autoplay delay milliseconds.                         |
| autoplayDisableOnInteraction |  true   | If set true, `autoplay` is disabled when use swipes. |

## Build in layouts

![](https://github.com/jzoom/images/raw/master/layout1.gif)

```dart
Swiper(
  itemBuilder: (BuildContext context, int index) {
    return Image.network(
      "https://via.placeholder.com/288x188",
      fit: BoxFit.fill,
    );
  },
  itemCount: 10,
  viewportFraction: 0.8,
  scale: 0.9,
)

```

![](https://github.com/jzoom/images/raw/master/layout2.gif)

```dart
Swiper(
  itemBuilder: (BuildContext context, int index) {
    return Image.network(
      "https://via.placeholder.com/288x188",
      fit: BoxFit.fill,
    );
  },
  itemCount: 10,
  itemWidth: 300.0,
  layout: SwiperLayout.STACK,
)
```

![](https://github.com/jzoom/images/raw/master/layout3.gif)

```dart
Swiper(
    itemBuilder: (BuildContext context, int index) {
      return Image.network(
        "https://via.placeholder.com/288x188",
        fit: BoxFit.fill,
      );
    },
    itemCount: 10,
    itemWidth: 300.0,
    itemHeight: 400.0,
    layout: SwiperLayout.TINDER,
 )
```

![](https://github.com/jzoom/images/raw/master/layout4.gif)

Very easy to create you own custom animation:

```dart

Swiper(
  layout: SwiperLayout.CUSTOM,
  customLayoutOption: CustomLayoutOption(
    startIndex: -1,
    stateCount: 3
  )..addRotate([
    -45.0/180,
    0.0,
    45.0/180
  ])..addTranslate([
    Offset(-370.0, -40.0),
    Offset(0.0, 0.0),
    Offset(370.0, -40.0)
  ]),
  itemWidth: 300.0,
  itemHeight: 200.0,
  itemBuilder: (context, index) {
    return Container(
      color: Colors.grey,
      child: Center(
        child: Text("$index"),
      ),
    );
  },
  itemCount: 10,
)

```

The `CustomLayoutOption` is designed to describe animations.
It is very easy to specify every state of items in Swiper.

```dart
CustomLayoutOption(
  // Which index is the first item of array below
  startIndex: -1,
  // array length
  stateCount: 3
)..addRotate([
  // rotation of every item
  -45.0/180,
  0.0,
  45.0/180
])..addTranslate([
  // offset of every item
  Offset(-370.0, -40.0),
  Offset(0.0, 0.0),
  Offset(370.0, -40.0)
]);

```

## Codes

![Example](https://github.com/jzoom/images/raw/master/swiper-example.gif)

```dart
ConstrainedBox(
  child: Swiper(
    outer:false,
    itemBuilder: (c, i) {
      return Wrap(
        runSpacing: 6.0,
        children: [0,1,2,3,4,5,6,7,8,9].map((i){
          return SizedBox(
            width: MediaQuery.of(context).size.width/5,
            child: Column(
              mainAxisSize: MainAxisSize.min,
              children: <Widget>[
                SizedBox(
                  child: Container(
                    child: Image.network("https://fuss10.elemecdn.com/c/db/d20d49e5029281b9b73db1c5ec6f9jpeg.jpeg%3FimageMogr/format/webp/thumbnail/!90x90r/gravity/Center/crop/90x90"),
                  ),
                  height: MediaQuery.of(context).size.width * 0.12,
                  width: MediaQuery.of(context).size.width * 0.12,
                ),
                Padding(padding: EdgeInsets.only(top:6.0),child: Text("$i"),)
              ],
            ),
          );
        }).toList(),
      );
    },
    pagination: SwiperPagination(
      margin: EdgeInsets.all(5.0)
    ),
    itemCount: 10,
  ),
    constraints:BoxConstraints.loose(Size(screenWidth, 170.0))
),

```

# FAQs

## How to remove default controlls?

- Do not pass any value in `control` parameter of constructor

## How to remove default pagination?

- Do not pass any value in `pagination` parameter of constructor

---

You can find all custom options here:

> https://github.com/TheAnkurPanchani/flutter_swiper/blob/master/example/lib/src/ExampleCustom.dart


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## [0.0.1] - 2023-05-03

- First release after fork


example/lib/src/example_custom.dart
import 'package:carousel_animations/carousel_animations.dart';
import 'package:flutter/material.dart';

import 'config.dart';
import 'forms/form_widget.dart';

class ExampleCustom extends StatefulWidget {
  const ExampleCustom({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return _ExampleCustomState();
  }
}

class _ExampleCustomState extends State<ExampleCustom> {
  //properties want to custom
  late int _itemCount;

  late bool _loop;

  late bool _autoplay;

  late int _autoplayDelay;

  late double _padding;

  late bool _outer;

  late double _radius;

  late double _viewportFraction;

  late SwiperLayout _layout;

  late int _currentIndex;

  late double _scale;

  late Axis _scrollDirection;

  late AxisDirection _axisDirection;

  late Curve _curve;

  late double _fade;

  late bool _autoplayDisableOnInteraction;

  late CustomLayoutOption customLayoutOption;

  late SwiperController _controller;

  TextEditingController numberController = TextEditingController();

  Widget _buildItem(BuildContext context, int index) {
    return ClipRRect(
      borderRadius: BorderRadius.all(Radius.circular(_radius)),
      child: Image.asset(
        images[index % images.length],
        fit: BoxFit.fill,
      ),
    );
  }

  @override
  void didUpdateWidget(ExampleCustom oldWidget) {
    customLayoutOption = CustomLayoutOption(startIndex: -1, stateCount: 3)
      ..addRotate([-45.0 / 180, 0.0, 45.0 / 180])
      ..addTranslate(
        [
          const Offset(-370.0, -40.0),
          Offset.zero,
          const Offset(370.0, -40.0),
        ],
      );
    super.didUpdateWidget(oldWidget);
  }

  @override
  void initState() {
    customLayoutOption = CustomLayoutOption(startIndex: -1, stateCount: 3)
      ..addRotate([-25.0 / 180, 0.0, 25.0 / 180])
      ..addTranslate(
        [
          const Offset(-350.0, 0.0),
          Offset.zero,
          const Offset(350.0, 0.0),
        ],
      );
    _fade = 1.0;
    _currentIndex = 0;
    _curve = Curves.ease;
    _scale = 0.8;
    _autoplay = false;
    _controller = SwiperController();
    _layout = SwiperLayout.TINDER;
    _radius = 10.0;
    _padding = 0.0;
    _loop = true;
    _itemCount = 3;
    _autoplayDelay = 3000;
    _viewportFraction = 0.8;
    _outer = false;
    _scrollDirection = Axis.horizontal;
    _axisDirection = AxisDirection.left;
    _autoplayDisableOnInteraction = false;
    super.initState();
  }

// maintain the index

  Widget buildSwiper() {
    return Swiper(
      onTap: (index) {
        Navigator.of(context).push(MaterialPageRoute<Object>(
          builder: (context) {
            return Scaffold(
              appBar: AppBar(
                title: const Text('New page'),
              ),
              body: Container(),
            );
          },
        ));
      },
      customLayoutOption: customLayoutOption,
      fade: _fade,
      index: _currentIndex,
      onIndexChanged: (index) {
        setState(() {
          _currentIndex = index;
        });
      },
      curve: _curve,
      scale: _scale,
      itemWidth: 300.0,
      controller: _controller,
      layout: _layout,
      outer: _outer,
      itemHeight: 200.0,
      viewportFraction: _viewportFraction,
      autoplayDelay: _autoplayDelay,
      loop: _loop,
      autoplay: _autoplay,
      itemBuilder: _buildItem,
      itemCount: _itemCount,
      scrollDirection: _scrollDirection,
      axisDirection: _axisDirection,
      indicatorLayout: PageIndicatorLayout.COLOR,
      autoplayDisableOnInteraction: _autoplayDisableOnInteraction,
      pagination: const SwiperPagination(
          builder: DotSwiperPaginationBuilder(
              size: 20.0, activeSize: 20.0, space: 10.0)),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        Container(
          color: Colors.black87,
          child: SizedBox(
            height: 300.0,
            width: double.infinity,
            child: buildSwiper(),
          ),
        ),
        Expanded(
          child: ListView(
            children: <Widget>[
              Text('Index:$_currentIndex'),
              Row(
                children: <Widget>[
                  ElevatedButton(
                    onPressed: () {
                      _controller.previous(animation: true);
                    },
                    child: const Text('Prev'),
                  ),
                  ElevatedButton(
                    onPressed: () {
                      _controller.next(animation: true);
                    },
                    child: const Text('Next'),
                  ),
                  Expanded(
                      child: TextField(
                    controller: numberController,
                  )),
                  ElevatedButton(
                    onPressed: () {
                      final text = numberController.text;
                      setState(() {
                        _currentIndex = int.parse(text);
                      });
                    },
                    child: const Text('Update'),
                  ),
                ],
              ),
              FormWidget(
                label: 'layout',
                child: FormSelect<SwiperLayout>(
                  placeholder: 'Select layout',
                  value: _layout,
                  values: const [
                    SwiperLayout.DEFAULT,
                    SwiperLayout.STACK,
                    SwiperLayout.TINDER,
                    SwiperLayout.CUSTOM
                  ],
                  valueChanged: (value) {
                    _layout = value;
                    setState(() {});
                  },
                ),
              ),
              FormWidget(
                label: 'scrollDirection',
                child: Switch(
                    value: _scrollDirection == Axis.horizontal,
                    onChanged: (value) => setState(() => _scrollDirection =
                        value ? Axis.horizontal : Axis.vertical)),
              ),
              if (_layout == SwiperLayout.STACK)
                FormWidget(
                  label: 'axisDirection (left <-> right)',
                  child: Switch(
                      value: _axisDirection == AxisDirection.right,
                      onChanged: (value) => setState(() => _axisDirection =
                          value ? AxisDirection.right : AxisDirection.left)),
                ),
              FormWidget(
                label: 'autoplayDisableOnInteraction',
                child: Switch(
                    value: _autoplayDisableOnInteraction,
                    onChanged: (value) =>
                        setState(() => _autoplayDisableOnInteraction = value)),
              ),
              //Pannel Begin
              FormWidget(
                label: 'loop',
                child: Switch(
                    value: _loop,
                    onChanged: (value) => setState(() => _loop = value)),
              ),
              FormWidget(
                label: 'outer',
                child: Switch(
                    value: _outer,
                    onChanged: (value) => setState(() => _outer = value)),
              ),
              //Pannel Begin
              FormWidget(
                label: 'autoplay',
                child: Switch(
                    value: _autoplay,
                    onChanged: (value) => setState(() => _autoplay = value)),
              ),

              FormWidget(
                label: 'padding',
                child: NumberPad(
                  number: _padding,
                  step: 5.0,
                  min: 0.0,
                  max: 30.0,
                  onChangeValue: (value) {
                    _padding = value.toDouble();
                    setState(() {});
                  },
                ),
              ),
              FormWidget(
                label: 'scale',
                child: NumberPad(
                  number: _scale,
                  step: 0.1,
                  min: 0.0,
                  max: 1.0,
                  onChangeValue: (value) {
                    _scale = value.toDouble();
                    setState(() {});
                  },
                ),
              ),
              FormWidget(
                label: 'fade',
                child: NumberPad(
                  number: _fade,
                  step: 0.1,
                  min: 0.0,
                  max: 1.0,
                  onChangeValue: (value) {
                    _fade = value.toDouble();
                    setState(() {});
                  },
                ),
              ),
              FormWidget(
                label: 'itemCount',
                child: NumberPad(
                  number: _itemCount,
                  step: 1,
                  min: 0,
                  max: 100,
                  onChangeValue: (value) {
                    _itemCount = value.toInt();
                    setState(() {});
                  },
                ),
              ),

              FormWidget(
                label: 'radius',
                child: NumberPad(
                  number: _radius,
                  step: 1.0,
                  min: 0.0,
                  max: 30.0,
                  onChangeValue: (value) {
                    _radius = value.toDouble();
                    setState(() {});
                  },
                ),
              ),

              FormWidget(
                label: 'viewportFraction',
                child: NumberPad(
                  number: _viewportFraction,
                  step: 0.1,
                  max: 1.0,
                  min: 0.5,
                  onChangeValue: (value) {
                    _viewportFraction = value.toDouble();
                    setState(() {});
                  },
                ),
              ),

              FormWidget(
                label: 'curve',
                child: FormSelect<Curve>(
                  placeholder: 'Select curve',
                  value: _curve,
                  values: const [
                    Curves.easeInOut,
                    Curves.ease,
                    Curves.bounceInOut,
                    Curves.bounceOut,
                    Curves.bounceIn,
                    Curves.fastOutSlowIn
                  ],
                  valueChanged: (value) {
                    _curve = value;
                    setState(() {});
                  },
                ),
              ),
            ],
          ),
        )
      ],
    );
  }
}
